### PR TITLE
Added --json flag to conan upload for output

### PR DESCRIPTION
**Pull request by danimtb**
- [x] Refer to the issue that supports this Pull Request: #1895 
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. Also adding a description of the changes in the ``changelog.rst`` file. https://github.com/conan-io/docs

### CODE REVIEW

File: conans/client/cmd/uploader.py [hunk 1]
```diff
@@ -54,21 +57,23 @@ def upload(self, reference_or_pattern, package_id=None, all_packages=None,
                     conanfile_path = self._client_cache.conanfile(conan_ref)
                     conan_file = load_conanfile_class(conanfile_path)
                 except NotFoundException:
-                    raise NotFoundException("There is no local conanfile exported as %s"
-                                            % str(conan_ref))
+                    raise ConanException(("There is no local conanfile exported as %s" %
```
[the remark this answers is not in the document]
Comment (danimtb) [reply conans/client/cmd/uploader.py]: Sorry, my mistake

File: conans/client/conan_api.py [hunk 2]
```diff
@@ -338,26 +337,29 @@ def get_test_conanfile_path(tf):
...
-                                  install_folder=None,  # Not output anything
-                                  manifest_folder=manifest_folder,
-                                  manifest_verify=manifest_verify,
-                                  manifest_interactive=manifest_interactive,
-                                  remote_name=remote,
-                                  profile=profile,
-                                  build_modes=build_modes,
-                                  update=update,
-                                  keep_build=keep_build)
+            ret = self._manager.install(reference=reference,
```
Comment (lasote) [inline conans/client/conan_api.py]: This is weird. What is returning the manager and the package tester? I think we don't want these returns, just always the recorder.get_info(). Right?

File: conans/client/conan_api.py [hunk 3]
```diff
@@ -338,26 +337,29 @@ def get_test_conanfile_path(tf):
...
+                                        manifest_folder=manifest_folder,
+                                        manifest_verify=manifest_verify,
+                                        manifest_interactive=manifest_interactive,
+                                        remote_name=remote,
+                                        profile=profile,
+                                        build_modes=build_modes,
+                                        update=update,
+                                        keep_build=keep_build)
+        if ret is None:
+            return self._recorder.get_info()
```
Comment (lasote) [inline conans/client/conan_api.py]: The `create` method returns always the recorder.

File: conans/client/conan_api.py [hunk 4]
```diff
@@ -443,13 +445,14 @@ def install_reference(self, reference, settings=None, options=None, env=None,
...
-                              generators=generators,
-                              install_reference=True)
+        ret = self._manager.install(reference=reference, install_folder=install_folder,
+                                    remote_name=remote, profile=profile, build_modes=build,
+                                    update=update, manifest_folder=manifest_folder,
+                                    manifest_verify=manifest_verify,
+                                    manifest_interactive=manifest_interactive,
+                                    generators=generators, install_reference=True)
+        if ret is None:
+            return self._recorder.get_info()
```
Comment (lasote) [inline conans/client/conan_api.py]: The export_pkg doesn't have return defined yet. Remove the all the ret stuff and do not return anything.

File: conans/client/conan_api.py [hunk 5]
```diff
@@ -467,17 +470,19 @@ def install(self, path="", settings=None, options=None, env=None,
...
+                                    profile=profile,
+                                    build_modes=build,
+                                    update=update,
+                                    manifest_folder=manifest_folder,
+                                    manifest_verify=manifest_verify,
+                                    manifest_interactive=manifest_interactive,
+                                    generators=generators,
+                                    no_imports=no_imports)
+        if ret is None:
+            return self._recorder.get_info()
```
Comment (lasote) [inline conans/client/conan_api.py]: Install always returns the info from the recorder, not the manager return.

File: conans/client/cmd/uploader.py [hunk 6]
```diff
@@ -29,8 +29,9 @@ def __init__(self, client_cache, user_io, remote_manager, registry):
 
     def upload(self, reference_or_pattern, package_id=None, all_packages=None,
                force=False, confirm=False, retry=0, retry_wait=0, skip_upload=False,
-               integrity_check=False, no_overwrite=None, remote_name=None):
+               integrity_check=False, no_overwrite=None, remote_name=None, recorder=None):
```
Comment (memsharded) [inline conans/client/cmd/uploader.py]: It is very weird to accept a recorder=None, and fail if it is not defined.
But leave it as-is, no big issue.

File: conans/client/conan_command_output.py [hunk 7]
```diff
@@ -47,8 +47,8 @@ def json_build_order(self, info, json_output, cwd):
                 json_output = os.path.join(cwd, json_output)
             save(json_output, json_str)
 
-    def json_install(self, info, json_output, cwd):
-        cwd = os.path.abspath(cwd or get_cwd())
+    def json_output(self, info, json_output, cwd):
+        cwd = os.path.abspath(cwd or os.getcwd())
```
Comment (memsharded) [inline conans/client/conan_command_output.py]: Please use ``get_cwd()``

File: conans/client/conan_api.py [hunk 8]
```diff
@@ -271,7 +264,9 @@ def test(self, path, reference, profile_name=None, settings=None, options=None,
         profile = profile_from_args(profile_name, settings, options, env, cwd,
                                     self._client_cache)
         reference = ConanFileReference.loads(reference)
-        pt = PackageTester(self._manager, self._user_io)
+        recorder = ActionRecorder()
```
Comment (memsharded) [inline conans/client/conan_api.py]: Really the ``new`` command needs recorder?

File: conans/client/conan_api.py [hunk 9]
```diff
@@ -419,72 +387,90 @@ def export_pkg(self, conanfile_path, name, channel, source_folder=None, build_fo
...
-        self._manager.install(reference=reference, install_folder=install_folder, remote_name=remote,
-                              profile=profile, build_modes=build, update=update,
-                              manifest_folder=manifest_folder,
-                              manifest_verify=manifest_verify,
-                              manifest_interactive=manifest_interactive,
-                              generators=generators,
-                              install_reference=True)
+        try:
+            recorder = ActionRecorder()
+            cwd = cwd or os.getcwd()
```
Comment (memsharded) [inline conans/client/conan_api.py]: No, please don't use ``os.getcwd()`` anymore, use ``get_cwd()``

File: conans/client/conan_api.py [hunk 10]
```diff
@@ -531,16 +517,22 @@ def info_build_order(self, reference, settings=None, options=None, env=None,
                          install_folder=None):
         reference, profile = self._info_get_profile(reference, install_folder, profile_name, settings,
                                                     options, env)
-        graph = self._manager.info_build_order(reference, profile, build_order, remote, check_updates)
+
+        recorder = ActionRecorder()
```
Comment (memsharded) [inline conans/client/conan_api.py]: It seems that methods not using json output/recorder, shouldn't need to instantiate it?

**Pull request merged**